bitcoind-regtest/src/Bitcoin/Core/Regtest/Generator.hs

```diff
@@ -1,47 +1,84 @@
+{-# LANGUAGE NamedFieldPuns #-}
 {-# LANGUAGE NumericUnderscores #-}
 {-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TupleSections #-}
 
 module Bitcoin.Core.Regtest.Generator (
     generateWithTransactions,
 ) where
 
+import Bitcoin.Core.RPC (
+    BitcoindClient,
+    ListUnspentOptions (ListUnspentOptions),
+ )
+import qualified Bitcoin.Core.RPC as RPC
+import Bitcoin.Core.Regtest.Framework (NodeHandle, oneBitcoin, runBitcoind)
+import Control.Arrow ((&&&))
 import Control.Concurrent (threadDelay)
+import Control.Concurrent.Async (async, link)
 import Control.Exception (throwIO)
-import Control.Monad (forever, replicateM)
+import Control.Monad (forever, replicateM, when, (>=>))
+import Control.Monad.IO.Class (MonadIO, liftIO)
+import Control.Monad.Trans.Class (lift)
+import Control.Monad.Trans.State.Strict (
+    StateT (StateT),
+    evalStateT,
+    gets,
+    mapStateT,
+    modify',
+ )
+import Data.ByteString (ByteString)
+import Data.Functor (void)
+import Data.List (find, uncons, unfoldr)
+import qualified Data.Map.Strict as Map
+import Data.Maybe (mapMaybe)
 import qualified Data.Serialize as S
+import Data.Text (Text)
 import Data.Word (Word64)
-import Haskoin.Block (BlockHeight, blockTxns)
+import Haskoin (
+    Script (Script),
+    TxHash,
+    TxIn (TxIn),
+ )
+import qualified Haskoin as H
+import Haskoin.Block (BlockHeight)
 import Haskoin.Transaction (
     OutPoint (..),
     Tx (..),
     TxOut (..),
-    txHash,
  )
 import Network.HTTP.Client (Manager)
 
-import qualified Bitcoin.Core.RPC as RPC
-import Bitcoin.Core.Regtest.Framework (NodeHandle, runBitcoind)
-import Control.Monad.IO.Class (liftIO)
-import Control.Monad.Trans.Class (lift)
-import Control.Monad.Trans.State.Strict (evalStateT, get, gets, modify, put)
-import Data.ByteString.Base64 (encodeBase64)
-import Data.Maybe (fromJust)
-import Data.Sequence (Seq ((:<|)), (|>))
-import Data.Text (Text)
+-- | How much to pay out on each faucet request
+faucetPayment :: Word64
+faucetPayment = oneBitcoin
 
-processCoinbase :: Tx -> (OutPoint, Word64)
-processCoinbase tx0 = (OutPoint (txHash tx0) 0, outValue . head $ txOut tx0)
+-- | If the series count exceeds this value, stop generating series
+maxSeries :: Int
+maxSeries = seriesLength * seriesCount
 
-data GeneratorState = GeneratorState
-    { generatorSpendableOutputs :: [(OutPoint, Word64)]
-    , generatorUnspentCoinbases :: Seq (OutPoint, Word64)
-    }
+-- | The number of series in a group
+seriesCount :: Int
+seriesCount = length (outputsPerSeries :: [Int])
 
-{- | Generate many transactions per block with a certain mean fee rate
+-- | How many transactions in a series to broadcast
+seriesLength :: Int
+seriesLength = 10
 
-FIXME For some reason @bitcoind@ refuses to sign transactions if we set
- @splitCount@ too high.  What's going on here?
--}
+-- | Initial amount per output
+amountPerOutput :: Integral a => a
```

Comment:
Is this numeric type abstraction necessary?

Reply:
The alternative is to use `fromIntegral` in a few places.

Reply:
I don't have a strong opinion about which approach is better: a more general type or using `fromIntegral`.

Reply:
What are the different types?

Reply:
I was wrong and it was possible to give the constants mono-types without ending up peppering the code with `fromIntegral`.